File: MATLAB/itd.m
function itd_samples = itd(theta, Fs)
% itd calculates the interaural time difference given some azimuth theta
%
% Inputs:
%   theta - Azimuth (in degrees)
%   Fs    - Sampling frequency (e.g. 44100)
%
% Output:
%   itd_samples - The number of samples representing the calculated ITD
%
% Example:
%     itd(30, 44100) -->  -12
%     itd(30, 21500) -->  -6
%     itd(190, 35)   -->  35
%
% Other m-files required: none
% Subfunctions: none
% MAT-files required: IRC_1020_C_HRIR.mat
%
% See also: localize
%
% Author: David Gurevich, Hons. BMath Computer Science
%         University of Waterloo, 2025
% Email Address: dgurevic (at) uwaterloo (dot) ca
% December 2020

% ------------------- BEGIN CODE -------------------

d = 0.0875; % radius of head (metres)
c = 343; % speed of sound (metres/second)

left_delay = 0;
right_delay = 0;

if theta == 360 || theta == 180
    left_delay = 0;
    right_delay = 0;
elseif 0 < theta  && theta < 180
    left_delay = Fs * (-d * sind(theta)) / c;
    right_delay = Fs* (d * deg2rad(theta)) / c;
elseif 180 < theta && theta < 360
    left_delay = Fs * (d * deg2rad(theta)) / c;
    right_delay = Fs * (-d * sind(theta)) / c;
end

itd_samples = round(left_delay - right_delay);
% -------------------- END CODE --------------------
end

File: CPP/src/itd.cpp
/*
    itd.cpp implements a function which calculates interaural time delay between left and right audio tracks

    Author: David Gurevich (dgurevic (at) uwaterloo (dot) ca)

    Copyright (c) David Gurevich 2021
 */

#include "../include/itd.h"

int calculate_delay(double theta, int sample_rate) {
    int left_delay = 0;
    int right_delay = 0;

    if (theta == 360 || theta == 180) {
        left_delay = 0;
        right_delay = 0;
    } else if (0 < theta && theta < 180) {
        left_delay = sample_rate * (-HEAD_RADIUS * sin(theta * M_PI / 180)) / SPEED_SOUND;
        right_delay = sample_rate * (HEAD_RADIUS * (theta * M_PI / 180)) / SPEED_SOUND;
    } else if (180 < theta && theta < 360) {
        left_delay = sample_rate * (HEAD_RADIUS * (theta * M_PI / 180)) / SPEED_SOUND;
        right_delay = sample_rate * (-HEAD_RADIUS * sin(theta * M_PI / 180)) / SPEED_SOUND;
    }

    int itd_samples = left_delay - right_delay;

    return itd_samples;
}
